Add an official Governance structure

odow opened the pull request:
This PR adds a formal governance structure for the JuMP project. 

A related document is the SciML governance page: https://sciml.ai/governance/

Before merging, this PR requires approval from:

- [x] @mlubin
- [x] @blegat
- [x] @joaquimg 
- [x] @joehuchette 
- [x] @odow
- [x] @juan-pablo-vielma
- [x] @chkwon
- [x] @iainnz (optional)
- [x] Consensus on monthly developer call

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,110 @@
+* Access to proprietary information of their employer that could potentially
+  leak into their work with the Project.
+
+All members of the Council shall disclose any conflict of interest they may
```
shall disclose to whom? to the public or to the other members of the council?

odow replied:
Just to the council?

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,110 @@
+
+### Conflict of interest
+
+It is expected that Steering Council members will be employed at a wide range of
```
steering committee or steering council?

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,110 @@
+
+All members of the Council shall disclose any conflict of interest they may
+have. Members with a conflict of interest in a particular issue may participate
+in Council discussions on that issue, but must recuse themselves from voting on
```
What does the council vote on? (Not described)

odow replied:
Changed to `recuse themselves from authorizing or approving any expenditures.`. It's really just the financial COI.

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,110 @@
+* Search the forum for similar questions that may have been already answered.
+* Read the first post of [PSA: make it easier to help you](https://discourse.julialang.org/t/psa-make-it-easier-to-help-you/14757/81).
+
+Remember to post a simplified version of your problem that demonstrates the
```
This discussion about discourse doesn't really fit in with the governance structure.

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,110 @@
+### Gitter chatroom
+
+The [Gitter channel](https://gitter.im/JuliaOpt/JuMP-dev) is for
+developer-focused chat about JuMP. If you are hacking on JuMP and have a
```
Also doesn't fit in with the governance structure, unless it's a statement like "the council will maintain a developer chat room ...".

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,110 @@
+
+### JuMP-dev workshops
+
+Approximately once per year, the JuMP community meets for a multi-day workshop,
```
For a governance doc, this should talk about how the workshops are organized, e.g., the council chooses a chair.

mlubin reviewed (commented):
It's good that we have a starting point now. But this is missing a discussion about how decisions (technical and otherwise) are ultimately made. BDFL or majority vote or consensus? May need to discuss in a call.

odow commented:
> But this is missing a discussion about how decisions (technical and otherwise) are ultimately made. BDFL or majority vote or consensus?

Consensus discussion on Github or a monthly developer calls. But we nominate you as BDFL with sole veto power? Agree probably needs to be discussed on call.

mlubin commented:
(Not yet commenting on the BDFL nomination.) Another important issue is allocation of commit bits and having a clear pathway for community members to become either part of the core team or an effective owner of a JuMP-dev repository (e.g., wrappers).

odow commented on pages/governance.md:
```diff
@@ -0,0 +1,179 @@
+### Repository Maintainer
+
+Repository maintainers have `Maintain` rights to an individual repository as an
+outside collaborator. This allows them to push code to a branch in the
```
This is a good reason for standardization. I took a look and the outside collaborators have a mix of `Write`, `Maintain`, and `Admin` rights.

odow replied:
Here's the breakdown: https://docs.github.com/en/free-pro-team@latest/github/setting-up-and-managing-organizations-and-teams/repository-permission-levels-for-an-organization#repository-access-for-each-permission-level

The short answer is yes.

The main difference between Maintain and Admin is the ability to edit settings (archive, default branches, change deploy keys), invite external collaborators, delete issues, etc.

I think Maintain is the one we want. We don't want external collaborators adding new collaborators, or deleting the repository, for example. 

For people transferring repositories, there is a trade-off: transfer to JuMP-dev and get structure, help with maintenance, etc, and loose some permissions such as the ability to add collaborators.

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,179 @@
+Core contributors are added as members of the [JuMP-dev Github organization] and
+have `Admin` permission to every repository.
+
+Community members can become core contributors by invite from the Steering
```
I would say that a new core contributor should be added by consensus of the current core contributors and notification to the steering council. I don't think the steering council needs to get involved in technical decisions unless they're strategically significant.

odow commented on pages/governance.md:
```diff
@@ -0,0 +1,187 @@
+* [Miles Lubin](https://github.com/mlubin)
+* [Benoît Legat](https://github.com/blegat)
+* [Joaquim Dias Garcia](https://github.com/joaquimg)
+* [Joey Huchette](https://github.com/joehuchette)
```
Is there a reason @joehuchette has `Member` permission in JuMP-dev and not `Owner`?

mlubin replied:
A related question: @joehuchette, would you prefer to be listed under core contributors or emeritus core contributors?

joehuchette replied:
A bit conflicted, mostly because "emeritus" suggests retiring from contributing in perpetuity, which is not my intention. I'll defer to the BDFL :)

mlubin replied:
The only mechanism defined for transitioning from being a core developer to an emeritus core developer is by request, so you'll remain a core developer until you ask not to be :)

joehuchette commented on pages/governance.md:
```diff
@@ -0,0 +1,191 @@
+
+A new core contributor may be added by consensus of the current core
+contributors and notification to the Steering Council. Before becoming a core
+contributor, it is expected that the community member has already recieved
```
*received

joehuchette commented on pages/governance.md:
```diff
@@ -0,0 +1,191 @@
+
+### Technical
+
+All technical decision and made via consensus of the core contributors and
```
*decisions are

mlubin reviewed (commented):
I would suggest a bit of a restructuring to clarify the roles of the core developers and steering council. Right now, core developers are a level 3 heading under "commit rights" while the steering council is a level 2 heading, which doesn't seem right. The distinction between what's a technical decision that falls to core developers and what's "other" that falls to the steering council is also ambiguous.

What do you think?

In my view, the steering council should be secondary to the core developers. In terms of decision making, the core developers have ultimate authority on the direction of the project (nothing can really happen without their agreement). The steering council is there to support the core developers, e.g., by handling finances and representing JuMP with NumFOCUS. The steering council can gain additional decision-making power if the core developers decide to delegate, e.g., choosing a chair for the JuMP-dev workshop.

odow commented:
> What do you think?

Agree. How's this now?

mlubin reviewed (commented):
I have more comments on the text, but first a larger refactoring suggestion. The decision-making process discussion is smeared across different sections, e.g., "Core contributors are the ultimate authority on the direction of the JuMP project." in the definition of core contributors.

What about:
1) Basic definition of the various entities, i.e., core contributors, emeritus core contributors, repository maintainers, steering council (interface with NumFOCUS), BDFL, and how membership in each group is determined.
2) The decision-making process
3) GitHub permissions for each group (if applicable). The GitHub permissions are there to enable people to carry out the decision-making process, so they're secondary in that respect.
4) Channels of community involvement

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,235 @@
+members, in consultation with the core contributors, will invite a member of the
+community to join in order to maintain a quorum of five members.
+
+#### Conflict of Interest
```
I would move "Conflict of interest" under the "decision making process" section and extend it to all decisions made by the steering council or core contributors. Core contributors should also disclose conflicts of interest with other core contributors and step back from decisions when conflicts of interests are in play.

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,235 @@
+spent in a manner that furthers the mission of JuMP. Fiscal decisions require
+majority approval by acting Steering Council members.
+
+The Steering Council can gain addition decision-making power if the core
```
This sentence is related to non-fiscal decisions.

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,235 @@
+This section outlines how fiscal and non-fiscal decisions are made in the JuMP
+project.
+
+### Fiscal Decisions
```
I think we mean financial more than fiscal.

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,235 @@
+
+All fiscal decisions are made by the Steering Council to ensure any funds are
+spent in a manner that furthers the mission of JuMP. Fiscal decisions require
+majority approval by acting Steering Council members.
```
Nit: "acting" is redundant.

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,235 @@
+
+## Github Permissions
+
+For security, all core contributors and repository maintainers must have
```
This is a setting that's enforced by GitHub, not sure how important it is to mention here. I would say that GitHub permissions are used to control access to repositories.

Also: Anyone with commit access to a repository is trusted to use it in a way that's consistent with the "non-fiscal decisions" process. Prefer pull requests over direct pushes, ask for feedback on changes if you're not sure there's a consensus, and follow JuMP's style guide and development processes.

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,235 @@
+
+### Emeritus Core Contributors
+
+Unless directed, emeritus core contributors retain the same commit rights as
```
"Emeritus core contributors retain the same commit rights as core contributors, unless they choose to surrender them."

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,235 @@
+
+### Community Forum
+
+The Steering Council will maintain a forum for community members to post
```
The steering council isn't doing much to maintain the forum. Do we need to assign this task to them, or just say that *there is* a community form? Same for the chat room.

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,235 @@
+
+### JuMP-dev Workshops
+
+The core contributors (or Steering Council, if delegated) will periodically
```
We can just choose to delegate this to the steering council permanently (don't see why we'd want to change it).

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,235 @@
+### Emeritus Core Contributors
+
+Emeritus core contributors are community members who were core contributors, but
+are no longer active participants in the JuMP project.
```
"... who were core contributors, but have stepped back to a less active role. A core developer may choose to switch to emeritus status by informing the other core developers and the steering council."

odow commented on pages/governance.md:
```diff
@@ -0,0 +1,250 @@
+
+Repository maintainers are trusted members of the community who help the core
+contributors by managing a limited number of repositories in the
+[JuMP-dev Github organization].
```
Do we need something about the core contributors may remove someone as a maintainer if they abuse their access, or if they are inactive for a period of time?

mlubin replied:
I'd say, "Repository maintainers can step back from their role at any time by informing a core developer. Furthermore, core developers. by consensus. can choose to remove repository maintainers for any reason (typically inactivity)."

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,276 @@
+the [JuMP-dev Github organization].
+
+Anyone with commit access to a repository is trusted to use it in a way that is
+consistent with the "non-financial decisions" process. Prefer pull requests over
```
For a consistent tone, "Those with permissions should prefer pull requests ..."

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,276 @@
+direct pushes, ask for feedback on changes if you are not sure there is a
+consensus, and follow JuMP's [style guide](https://jump.dev/JuMP.jl/stable/style/)
+and development processes. Core contributors and repository maintainers who
+abuse their commit access with have their permissions revoked.
```
I don't think this threatening sentence is needed. From my suggestion above, core developers would be able to remove repository maintainers for any reason, which would include abuse. More extreme cases are also covered by the code of conduct and conflict of interest sections.

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,281 @@
+
+Community members wishing to transfer a repository from their personal account
+to the [JuMP-dev Github organization] should contact the [core contributors]
+via the [developer chatroom].
```
Maybe this is a good place to mention that we request you not use "JuMP" in the name of third-party packages.

mlubin commented:
Now's a good time to get feedback from the other core developers. @blegat @joaquimg

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,283 @@
+
+Community members may be maintainers of multiple repositories at the same time.
+
+Each repository should list the current maintainers in their corresponding
```
This might create an artificial barrier. Some people might be good repository maintainers but not want to have their name on the README, which potentially invites attention and demands from users. We could use a separate and less user-facing page to list repository maintainers.

odow replied:
Fair point. I've removed for now. We can revisit if/how to give maintainers recognition in future. Are you thinking a `MAINTAINERS.md` in each repo? Or a page on `jump.dev`?

mlubin replied:
I was thinking a page on jump.dev.

joehuchette commented on pages/governance.md:
```diff
@@ -0,0 +1,280 @@
+
+## Mission
+
+The mission of the JuMP project is to provide a free, open-source, and
```
I would propose "powerful" and "easy-to-use" as other potential descriptors.

joehuchette commented on pages/governance.md:
```diff
@@ -0,0 +1,280 @@
+This section outlines the different entities that exist within the JuMP project,
+their basic role, and how membership of each entity is determined.
+
+### BDFL
```
Worth spelling out what this is an acronym for at least once, somewhere.

joehuchette commented on pages/governance.md:
```diff
@@ -0,0 +1,280 @@
+contributors and notification to the [Steering Council].
+
+Before becoming a core contributor, it is expected that the community member is
+a [repository maintainer](#repository-maintainers) of several repositories.
```
several repositories under the jump-dev umbrella?

joaquimg commented on pages/governance.md:
```diff
@@ -0,0 +1,280 @@
+
+All non-financial decisions are made via consensus of the [core contributors]
+and relevant [repository maintainers]. [Emeritus core contributors](#emeritus-core-contributors)
+ are not considered to be [core contributors] for this purpose.
```
remove space

joaquimg commented on pages/governance.md:
```diff
@@ -0,0 +1,280 @@
+
+Community members proposing decisions with a financial aspect should contact the
+[Steering Council] directly, or table their proposal as an agenda item for
+discussion on a [monthly developer call](#monthly-developer-call).
```
make it clearer how proposals are tabled, same for non-financial.
possibly this additional detail could live in: ### Monthly Developer Call
and could be as simple as ping mlubin or odow

joaquimg commented on pages/governance.md:
```diff
@@ -0,0 +1,280 @@
+[Bordeaux, 2018](https://jump.dev/meetings/bordeaux2018/), and [Boston, 2017](https://jump.dev/meetings/mit2017/).
+
+Announcements of future workshops will be communicated through the
+[community forum] and via the [jump.dev](https://jump.dev) website.
```
Should we include jump's twitter here?

joaquimg reviewed (approved):
Great work!
Left some comments.

odow commented on pages/governance.md:
```diff
@@ -0,0 +1,280 @@
+the [NumFOCUS Code of Conduct](https://numfocus.org/code-of-conduct) and the
+[Julia Community Standards](https://julialang.org/community/standards/), which
+reflect the values of our community. Both these links have information on how to
+confidentially report a violation.
```
Once this governance document is finalized, I propose we replace the `CODE_OF_CONDUCT.md` with a link here instead.

blegat commented on _posts/2020-10-20-governance.md:
```diff
@@ -0,0 +1,39 @@
+ * Formalizing a process for transferring repositories to the
+   [jump-dev](https://github.com/jump-dev) Github organization.
+
+We belive this document strengthens the foundation of JuMP, and sets it up for
```
belive -> believe

blegat reviewed (approved):
Looks good!

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,291 @@
+fiscal, legal, and administrative support to help ensure the health and
+sustainability of the project.
+
+### Steering Council
```
The name of the list is jump-steering-committee@...
Did we intend to rename the committee to a council?

odow replied:
> Did we intend to rename the committee to a council?

I copied the SciML language, and couldn't remember if it was a council or a committee. Will change.

chkwon commented on pages/governance.md:
```diff
@@ -0,0 +1,291 @@
+invited by a [core contributor](#core-contributors) to present their proposal at
+a [monthly developer call] for further discussion and community input.
+
+Non-code-related decisions, such long-term strategic planning for JuMP, should
```
such as

chkwon commented on pages/governance.md:
```diff
@@ -0,0 +1,291 @@
+If consensus on a non-financial decision cannot be achieved, the final decision
+will be made by the [BDFL](#benevolent-dictator-for-life).
+
+The [Steering Committee] can gain addition decision-making power if the [core
```
additional

chkwon commented on pages/governance.md:
```diff
@@ -0,0 +1,291 @@
+of the JuMP documentation.
+
+Note that when developing a new solver or JuMP-extension, the [core
+contributors] request that you not use "JuMP" in the name without prior consent.
```
do not use

chkwon commented on pages/governance.md:
```diff
@@ -0,0 +1,291 @@
+Even once a package has matured, the bar for transferring repositories to the
+[JuMP-dev Github organization] is high, and can be summarized by the following:
+
+* The maintainance of the package would be simplified if [core contributors]
```
maintenance

chkwon reviewed (commented):
Looks good to me.

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,291 @@
+the [JuMP-dev Github organization].
+
+Anyone with commit access to a repository is trusted to use it in a way that is
+consistent with the [non-financial decisions](#non-financial-decisions) process.
```
It's fine to say "consistent with the Decision Making Process", because commit access might be used for financial-related decisions also.

mlubin commented on pages/governance.md:
```diff
@@ -0,0 +1,292 @@
+### Benevolent Dictator For Life
+
+Due to his role in the creation of JuMP, [Miles Lubin](https://github.com/mlubin)
+holds the role of [Benevolent Dictator For Life (BDFL)](https://en.wikipedia.org/wiki/Benevolent_dictator_for_life).
```
Nit: "holds the title of" sounds better

odow replied:
Fixed, boss.

joehuchette reviewed (approved):
LGTM

juan-pablo-vielma commented:
LGTM

IainNZ commented:
LGTM

odow commented:
I'll leave this open for the weekend if anyone has additional comments, particularly the new wording of the mission. If there are no issues raised, I will merge Sunday evening US Eastern time.

joaquimg commented:
One minor, non-blocking, comment.
Did we put the governance tab as the second tab on the webpage on purpose?

In https://sciml.ai/governance/ it is close to the end...
In https://scikit-learn.org/stable/about.html is it even a bit hidden in "about us".
A bit hidden in https://docs.scipy.org/doc/scipy/reference/index.html also...

Not suggesting it should be hidden, but I would move it to the end.

odow commented:
We now have:
<img width="692" alt="image" src="https://user-images.githubusercontent.com/8177701/97052890-5355ff80-15de-11eb-9a62-22493201f392.png">
I moved the forum, chatroom, and Code/Ecosystem page under "Community", and renamed "News" to "Blog"

mlubin commented on _config.yml:
```diff
@@ -111,10 +111,16 @@ css_inline: false # Will insert all styles into a single <style> block in the <h
+  items:
+  - title: Forum
+    url: https://discourse.julialang.org/c/domain/opt
+  - title: Chatroom
```
This is a developer chatroom. We shouldn't be funneling users into it by advertising it as the JuMP chatroom.
